lib/TPP/Transforms/HoistVectorTransfers.cpp

```diff
@@ -0,0 +1,259 @@
+  auto subviewOpRhsOffsets = subviews[1].getOffsets();
+  auto subviewOpAccOffsets = subviews[2].getOffsets();
+
+  SmallVector<scf::ForOp> list;
+  scf::ForOp current = contractOp->getParentOfType<scf::ForOp>();
+  if (!current)
+    return failure();
+
+  // check the induction variable usage in subviews of vector read op
+  Value ivK = current.getInductionVar();
+  if (ivK != subviewOpLhsOffsets[2] || ivK != subviewOpRhsOffsets[1])
+    return failure();
+
+  list.push_back(current);
+  for (int i = 0; i < 3; i++) {
+    scf::ForOp parent = current->getParentOfType<scf::ForOp>();
+    if (!parent)
+      return failure();
+    list.push_back(parent);
+    current = parent;
+  }
+
+  // Check the induction variable usgae in subviews of vector read op for other
+  // loops
+  Value ivReduction = list[1].getInductionVar();
+  if (ivReduction != subviewOpLhsOffsets[0] ||
+      ivReduction != subviewOpRhsOffsets[0])
+    return failure();
+
+  Value ivN = list[2].getInductionVar();
+  if (ivN != subviewOpAccOffsets[1] || ivN != subviewOpRhsOffsets[2])
+    return failure();
+
+  Value ivM = list[3].getInductionVar();
+  if (ivM != subviewOpLhsOffsets[1] || ivM != subviewOpAccOffsets[0])
+    return failure();
+
+  return list;
+}
+
+struct HoistVectorTransferOp : OpRewritePattern<vector::ContractionOp> {
+  using OpRewritePattern<vector::ContractionOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::ContractionOp contractOp,
+                                PatternRewriter &rewriter) const override {
+
+    // Check the vector contract operation satisfies the required pattern.
+    // Check the Acc, Lhs, and Rhs of contract operation
+
+    auto operands = getContractOperands(contractOp);
+    if (failed(operands))
+      return rewriter.notifyMatchFailure(contractOp,
+                                         "Invalid operands for contract op");
+
+    auto readOp = *operands;
+    auto vectorReadOpAcc = readOp[2];
+    auto vectorReadOpLhs = readOp[0];
+    auto vectorReadOpRhs = readOp[1];
+
+    // Check whether the operand of vector transfer read is a subview
+    auto subviews = getReadOperands(readOp);
+    if (failed(subviews))
+      return rewriter.notifyMatchFailure(
+          contractOp, "Vector read op operand(s) is/are not a subview");
+
+    // Check the operation type MatMul, B-MatMul, or BR-MatMul
+    auto contractIteratorTypes = contractOp.getIteratorTypesArray();
+    size_t reductionCount = 0;
+    for (size_t i = 0; i < contractIteratorTypes.size(); i++) {
+      if (contractIteratorTypes[i] == vector::IteratorType::reduction)
+        reductionCount++;
+    }
+
+    auto vectorReadOpLhsType = cast<ShapedType>(vectorReadOpLhs.getType());
+    auto vectorReadOpRhsRank =
+        (cast<ShapedType>(vectorReadOpRhs.getType())).getRank();
+
+    if (reductionCount == 2 &&
+        (vectorReadOpLhsType.getRank() != 3 || vectorReadOpRhsRank != 3))
+      return failure();
```

Comment:
Couldn't this be a `notifyMatchFailure` to?

Reply:
Okay. Replaced with `notifyMatchFailure `